### PR TITLE
Do not panic on missing close paren

### LINKED ISSUE #58856: Missing right parenthesis panics the compiler
I wrote some shitty code here: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d44eb1015f294e937d86b157339d0978
And it broke the compiler. Apparently by reaching supposedly unreachable code of a parsing function.

Some background can be found at https://users.rust-lang.org/t/the-squeeze-challenge/25762/10
As pointed by @cuviper the problem can be reduced to
```rust
trait Howness {}
impl Howness for (){
    fn how_are_you(&self->Empty
    {
        Empty
    }
}
```

### PR DESCRIPTION
**Pull request by estebank**
Fix #58856.

### CODE REVIEW

File: src/test/ui/issues/issue-58856-1.stderr [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+   |
+LL |     fn b(self>
+   |              ^ expected one of `->`, `where`, or `{` here
+
+error: expected one of `->`, `async`, `const`, `crate`, `default`, `existential`, `extern`, `fn`, `pub`, `type`, `unsafe`, `where`, or `}`, found `>`
+  --> $DIR/issue-58856-1.rs:3:14
+   |
+LL |     fn b(self>
+   |              ^ expected one of 13 possible tokens here
+
```
Comment (estebank) [inline src/test/ui/issues/issue-58856-1.stderr]: The two extra errors here are less than ideal, but this is something that can be ameliorated by keeping track of unexpected errors and elide the later ones based on the primary span.

Comment (jethrogb) [reply src/test/ui/issues/issue-58856-1.stderr]: I think you can get rid of them by adding `{}` after `>`

Comment (estebank) [reply src/test/ui/issues/issue-58856-1.stderr]: @jethrogb I can for this test, but I would like to get rid of them in a more general way to account for this kind of code in the wild.

**Review by hdhoang — COMMENTED** [review]
very sorry about the mis-comment, I will try to revoke this

**Comment by estebank** [comment]
Added to #58903.

**Pull request closed without merging**